ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,340 @@
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+// ReadCheckpointV6 reads checkpoint file from a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+func ReadCheckpointV6(dir string, fileName string) ([]*trie.MTrie, error) {
+	// TODO: read the main file and check the version
+	// TODO: read the checksum of 17 file parts
+
+	headerPath := filePathHeader(dir, fileName)
+	subtrieChecksums, topTrieChecksum, err := readHeader(headerPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	subtrieNodes, err := readSubTriesConcurrently(dir, fileName, subtrieChecksums)
+	if err != nil {
+		return nil, fmt.Errorf("could not read subtrie from dir: %w", err)
+	}
+
+	// validate the subtrie node checksum
+	tries, err := readTopLevelTries(dir, fileName, subtrieNodes, topTrieChecksum)
+	if err != nil {
+		return nil, fmt.Errorf("could not read top level nodes or tries: %w", err)
+	}
+
+	return tries, nil
+}
+
+func filePathHeader(dir string, fileName string) string {
+	return path.Join(dir, fileName)
+}
+
+func filePathSubTries(dir string, fileName string, index int) (string, string, error) {
+	if index < 0 || index > 15 {
+		return "", "", fmt.Errorf("index must be between 1 to 16, but got %v", index)
+	}
+	subTrieFileName := fmt.Sprintf("%v.%v", fileName, index)
+	return path.Join(dir, subTrieFileName), subTrieFileName, nil
+}
+
+func filePathTopTries(dir string, fileName string) (string, string) {
+	topTriesFileName := fmt.Sprintf("%v.%v", fileName, 16)
+	return path.Join(dir, topTriesFileName), fileName
+}
+
+// TODO: to add
+func readHeader(filePath string) ([]uint32, uint32, error) {
+	return make([]uint32, 16), 0, nil
+}
+
+// func readNodeCountAndTriesCount(f *os.File) (uint64, uint16, error) {
+// 	// read the node count and tries count from the end of the file
+// 	const footerOffset = encNodeCountSize + encTrieCountSize + crc32SumSize
+// 	const footerSize = encNodeCountSize + encTrieCountSize // footer doesn't include crc32 sum
+// 	footer := make([]byte, footerSize)                     // must not be less than 1024
+// 	_, err := f.Seek(-footerOffset, io.SeekEnd)
+// 	if err != nil {
+// 		return 0, 0, fmt.Errorf("cannot seek to footer: %w", err)
+// 	}
+// 	_, err = io.ReadFull(f, footer)
+// 	if err != nil {
+// 		return 0, 0, fmt.Errorf("could not read footer: %w", err)
+// 	}
+//
+// 	return decodeFooter(footer)
+// }
+
+func readSubTriesConcurrently(dir string, fileName string, subtrieChecksums []uint32) ([][]*node.Node, error) {
+	// TODO: replace 16 with const
+	if len(subtrieChecksums) != 16 {
+		return nil, fmt.Errorf("expect subtrieChecksums to be %v, but got %v", 16, len(subtrieChecksums))
+	}
```

Comment:
I have added [the check you mentioned](https://github.com/onflow/flow-go/pull/3321/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR37). 

However, I still think it's good to keep this check here as a sanity check. Since this function returns an error anyway, removing this check wouldn't save us too much.